### PR TITLE
[DPE-4066] HA interface

### PR DESCRIPTION
**Pull request by dragomirp**
* Adds hacluster interface 
* Adds a vip config option
* Updates charm libs

Related to #212

### CODE REVIEW

**Comment by gustavosr98** [comment]
Hey @dragomirp (cc @taurus-forever)

I tried the bundle you send to me
With Juju 3.4.5 on top of LXD 5.21

```yaml
description: PGB VIP bundle
name: vip-bundle
series: jammy
applications:
  data-integrator:
    charm: data-integrator
    channel: latest/stable
    num_units: 3
    options:
      database-name: testdb
  postgresql:
    charm: postgresql
    channel: 14/stable
    num_units: 3
    #options:
    #  profile: testing
  pgbouncer:
    channel: 1/edge/hacluster
    revision: 347
    charm: pgbouncer
    series: jammy
    num_units: 0
    options:
      max_db_connections: 30
      vip: 1
relations:
    - [pgbouncer:backend-database, postgresql:database]
    - [pgbouncer:database, data-integrator:postgresql]
```

```bash
# blue@blue:~$ juju status
Model   Controller  Cloud/Region         Version  SLA          Timestamp
pg-vip  overlord    localhost/localhost  3.4.5    unsupported  23:39:19Z

App              Version  Status  Scale  Charm            Channel           Rev  Exposed  Message
data-integrator           active      3  data-integrator  latest/stable      27  no
pgbouncer        1.21.0   active      3  pgbouncer        1/edge/hacluster  347  no
postgresql       14.11    active      3  postgresql       14/stable         429  no

Unit                Workload  Agent  Machine  Public address  Ports     Message
data-integrator/0*  active    idle   0        10.58.159.146
  pgbouncer/0*      active    idle            10.58.159.146   6432/tcp
data-integrator/1   active    idle   1        10.58.159.150
  pgbouncer/1       active    idle            10.58.159.150   6432/tcp
data-integrator/2   active    idle   2        10.58.159.212
  pgbouncer/2       active    idle            10.58.159.212   6432/tcp
postgresql/0        active    idle   3        10.58.159.13    5432/tcp
postgresql/1*       active    idle   4        10.58.159.203   5432/tcp
postgresql/2        active    idle   5        10.58.159.20    5432/tcp  Primary

Machine  State    Address        Inst id        Base          AZ  Message
0        started  10.58.159.146  juju-fc0f0d-0  ubuntu@22.04      Running
1        started  10.58.159.150  juju-fc0f0d-1  ubuntu@22.04      Running
2        started  10.58.159.212  juju-fc0f0d-2  ubuntu@22.04      Running
3        started  10.58.159.13   juju-fc0f0d-3  ubuntu@22.04      Running
4        started  10.58.159.203  juju-fc0f0d-4  ubuntu@22.04      Running
5        started  10.58.159.20   juju-fc0f0d-5  ubuntu@22.04      Running
```

But It does not seem to be assigning the VIP to any machine
```
# blue@blue:~$ juju exec -a  data-integrator -- 'ip -o -c -4 a'
data-integrator/0:
1: lo    inet 127.0.0.1/8 scope host lo\       valid_lft forever preferred_lft forever
9: eth0    inet 10.58.159.146/24 metric 100 brd 10.58.159.255 scope global dynamic eth0\       valid_lft 3140sec preferred_lft 3140sec

data-integrator/1:
1: lo    inet 127.0.0.1/8 scope host lo\       valid_lft forever preferred_lft forever
11: eth0    inet 10.58.159.150/24 metric 100 brd 10.58.159.255 scope global dynamic eth0\       valid_lft 3273sec preferred_lft 3273sec

data-integrator/2:
1: lo    inet 127.0.0.1/8 scope host lo\       valid_lft forever preferred_lft forever
13: eth0    inet 10.58.159.212/24 metric 100 brd 10.58.159.255 scope global dynamic eth0\       valid_lft 3128sec preferred_lft 3128sec

# blue@blue:~$ juju exec -a  data-integrator -- 'sudo crm_mon'
data-integrator/0:
sudo: crm_mon: command not found

data-integrator/1:
sudo: crm_mon: command not found

data-integrator/2:
sudo: crm_mon: command not found

ERROR the following tasks failed:
 - id "6" with return code 1
 - id "7" with return code 1
 - id "8" with return code 1

use 'juju show-task' to inspect the failures
```

**Comment by gustavosr98** [comment]
Adding some more 

```bash
juju deploy hacluster --channel 2.4/stable
juju relate hacluster data-integrator
juju relate hacluster:juju-info data-integrator:juju-info
```

Ok now it looks it is there 

```bash
# blue@blue:~$ ping 10.58.159.250
PING 10.58.159.250 (10.58.159.250) 56(84) bytes of data.
64 bytes from 10.58.159.250: icmp_seq=1 ttl=64 time=575 ms
64 bytes from 10.58.159.250: icmp_seq=2 ttl=64 time=0.099 ms
64 bytes from 10.58.159.250: icmp_seq=3 ttl=64 time=0.083 ms
^C
--- 10.58.159.250 ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2006ms
rtt min/avg/max/mdev = 0.083/191.571/574.533/270.794 ms
blue@blue:~$ juju exec -a  data-integrator -- 'ip -o -c -4 a'
data-integrator/0:
1: lo    inet 127.0.0.1/8 scope host lo\       valid_lft forever preferred_lft forever
9: eth0    inet 10.58.159.146/24 metric 100 brd 10.58.159.255 scope global dynamic eth0\       valid_lft 2195sec preferred_lft 2195sec
9: eth0    inet 10.58.159.250/24 scope global secondary eth0\       valid_lft forever preferred_lft forever

data-integrator/1:
1: lo    inet 127.0.0.1/8 scope host lo\       valid_lft forever preferred_lft forever
11: eth0    inet 10.58.159.150/24 metric 100 brd 10.58.159.255 scope global dynamic eth0\       valid_lft 2327sec preferred_lft 2327sec

data-integrator/2:
1: lo    inet 127.0.0.1/8 scope host lo\       valid_lft forever preferred_lft forever
13: eth0    inet 10.58.159.212/24 metric 100 brd 10.58.159.255 scope global dynamic eth0\       valid_lft 2183sec preferred_lft 2183sec
```

**Comment by gustavosr98** [comment]
Looks working

```bash 
# juju run data-integrator/leader get-credentials
Running operation 19 with 1 task
  - task 20 on unit-data-integrator-0

Waiting for task 20...
ok: "True"
postgresql:
  data: '{"database": "testdb", "external-node-connectivity": "true", "requested-secrets":
    "[\"username\", \"password\", \"tls\", \"tls-ca\", \"uris\"]"}'
  database: testdb
  endpoints: 10.58.159.250:6432
  password: REDACTED
  subordinated: "true"
  username: relation_id_7
  version: "14.11"

# psql --host 10.58.159.250 --port 6432 --username relation_id_7  --password --database testdb
psql (14.12 (Ubuntu 14.12-0ubuntu0.22.04.1), server 14.11 (Ubuntu 14.11-0ubuntu0.22.04.1))
Type "help" for help.
testdb=>
```

**Comment by dragomirp** [comment]
> Ok now it looks it is there

Hi, @gustavosr98, yes on LXD you need hacluster to get the vip. Adding an LXD bundle here for completeness:
```
description: PGB VIP bundle LXD
name: vip-bundle
series: jammy
applications:
  data-integrator:
    charm: data-integrator
    channel: latest/stable
    num_units: 3
    options:
      database-name: testdb
  postgresql:
    charm: postgresql
    channel: 14/stable
    num_units: 3
    #options:
    #  profile: testing
  pgbouncer:
    channel: 1/edge/hacluster
    revision: 347
    charm: pgbouncer
    series: jammy
    num_units: 0
    options:
      max_db_connections: 30
      #vip: "1.2.3.4"
  hacluster:
    charm: hacluster
    channel: 2.4/stable
    num_units: 0
relations:
    - [pgbouncer:backend-database, postgresql:database]
    - [pgbouncer:database, data-integrator:postgresql]
    - [hacluster:juju-info, data-integrator:juju-info]
    - [hacluster:ha, pgbouncer:ha]

```

**Comment by taurus-forever** [comment]
Hi @gustavosr98 , thank you for the instant testing here!
I have tested it from my side as well, and it works.
The only noticed error has been reported hacluster: https://bugs.launchpad.net/charm-hacluster/+bug/2076410

Great work @dragomirp !

File: src/charm.py [hunk 1]
```diff
@@ -910,14 +939,13 @@ def unit_ip(self) -> str:
     #  Relation Utilities
     # =====================
 
-    def update_client_connection_info(self, port: Optional[str] = None):
+    def update_client_connection_info(self):
```
Comment (dragomirp) [inline src/charm.py:942]: Not really passed anywhere

File: src/relations/hacluster.py
Comment (dragomirp) [inline src/relations/hacluster.py:1]: Mostly following the implementation in https://github.com/openstack/charm-interface-hacluster

Comment (dragomirp) [inline src/relations/hacluster.py:1]: Mostly following the implementation in https://github.com/openstack/charm-interface-hacluster

Comment (dragomirp) [inline src/relations/hacluster.py:1]: Mostly following the implementation in https://github.com/openstack/charm-interface-hacluster

File: src/charm.py [hunk 2]
```diff
@@ -507,8 +510,24 @@ def update_status(self):
...
+            return
+
+        if vip and not self._is_exposed:
+            self.unit.status = BlockedStatus("vip configuration without data-intgrator")
+            return
+
         if self.check_pgb_running():
-            self.unit.status = ActiveStatus()
+            if self.unit.is_leader() and vip:
+                self.unit.status = ActiveStatus(f"VIP: {vip}")
```
Comment (dragomirp) [inline src/charm.py:528]: Setting just the app status message here didn't show when testing locally, while setting a proper app status didn't clear up when blocking.

File: src/charm.py [hunk 3]
```diff
@@ -534,6 +557,10 @@ def _on_config_changed(self, event) -> None:
             except ModelError:
                 logger.exception("failed to open port")
 
+        if self.unit.is_leader():
+            self.peers.app_databag["current_port"] = str(self.config["listen_port"])
+            self.peers.app_databag["current_vip"] = str(vip)
```
Comment (marceloneppel) [inline src/charm.py:562]: When `vip` is not set, does this translate to `none` (lowercase) to perform the comparison from L544-545 correctly?

**Review by marceloneppel — APPROVED** [review]

File: src/charm.py (hunk 3 above)
Comment (dragomirp) [reply src/charm.py:562]: Should be handled on lines 543-544

Comment (dragomirp) [reply src/charm.py]: Cleared up the logic, so we don't have to rely on flags.

Comment (marceloneppel) [reply src/charm.py]: Ah, I missed that. Sorry.

**Review by marceloneppel — APPROVED** [review]

**Review by taurus-forever — APPROVED** [review]
Tested with HAcluster from stable. Work well. 
Also tested by Gustavo to ensure OpenStack case works the same as legacy pgbouncer.

Adding Mohamed to approve the config name.

**Pull request merged**